kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+        let tracked mut next_ptr = next_points_to.get();
+        let next_node = node.next(Tracked(&next_ptr));
+        next_node.0.write(Tracked(&mut next_ptr), old_head_value);
+        self.head.0.write(Tracked(&mut head_points_to), Some(node));
+        let cell = &next_node.0;
+
+        proof {
+            ghost_state.borrow_mut().cells = ghost_state@.cells.insert(1, cell);
+            ghost_state.borrow_mut().points_to_map.tracked_map_keys_in_place(
+                Map::<nat, nat>::new(
+                    |j: nat| 2 <= j && j < ghost_state@.cells.len(),
+                    |j: nat| (j - 1) as nat,
+                ),
+            );
+            ghost_state.borrow_mut().points_to_map.tracked_insert(0, head_points_to);
+            ghost_state.borrow_mut().points_to_map.tracked_insert(1, next_ptr);
+            assert forall|i: nat| 2 <= i < ghost_state@.cells.len() implies well_formed_node(
+                ghost_state,
+                i,
+            ) by {
+                assert(1 <= i - 1 < old(ghost_state)@.cells.len());
+                assert(old(ghost_state)@.cells[i - 1] == ghost_state@.cells[i as int]);
+                assert(old(ghost_state)@.points_to_map.dom().contains((i - 1) as nat)
+                    == ghost_state@.points_to_map.dom().contains(i));
+                assert(old(ghost_state)@.points_to_map[(i - 1) as nat]
+                    == ghost_state@.points_to_map[i]);
+            }
+        }
+    }
+
+    pub fn push_tail(
+        &'a self,
+        node: &'a T,
+        next_points_to: Tracked<PointsTo<Option<&'a T>>>,
+        ghost_state: &mut Tracked<GhostState<'a, T>>,
+    )
+        requires
+            old(ghost_state)@.cells.len() >= 1,
+            forall|i: nat|
+                0 <= i < old(ghost_state)@.cells.len() ==> #[trigger] old(
+                    ghost_state,
+                )@.points_to_map.dom().contains(i) && old(ghost_state)@.points_to_map[i].is_init()
+                    && old(ghost_state)@.points_to_map[i].id() == old(
+                    ghost_state,
+                )@.cells[i as int].id(),
+            forall|i: nat|
+                0 <= i < (old(ghost_state)@.cells.len() - 1) as nat ==> match #[trigger] old(
+                    ghost_state,
+                )@.points_to_map[i].value() {
+                    Option::Some(_) => true,
+                    Option::None => false,
+                },
+            match old(ghost_state)@.points_to_map[(old(ghost_state)@.cells.len()
+                - 1) as nat].value() {
+                Option::Some(_) => false,
+                Option::None => true,
+            },
+            old(ghost_state)@.cells[0].id() == self.head.0.id(),
+            next_points_to@.is_init(),
+        ensures
+            old(ghost_state)@.cells.len() + 1 == ghost_state@.cells.len(),
+            self.well_formed_list(ghost_state),
+            ghost_state@.points_to_map[(ghost_state@.cells.len() - 2) as nat].value().unwrap()
```

Comment:
Why is it `-2` and not `-1` ? Can you add a comment to explain why?

Reply:
Because the last cell in `ghost_state@.cells` is always `None` inside to indicate the end of list. It has to be included in verification due to how the `ListLinkV` struct and the `ListNodeV` trait are defined.